### PR TITLE
feat: Build for Python 3.14

### LINKED ISSUE #372: Add support for Python 3.14
Once support has been added to `pymgclient` (see https://github.com/memgraph/pymgclient/issues/81), we can also try adding it here.

### PR DESCRIPTION
**Pull request by mattkjames7**
### Description

This PR updates `gqlalchemy` for Python 3.14 readiness by combining CI/dependency compatibility updates with the previously merged Pydantic v2 API migration #384. It broadens test coverage across newer Python versions, adjusts optional dependency markers for unsupported stacks (notably TensorFlow/TF-GNN on newer interpreters), and keeps OGM behavior stable for normal `Node`/`Relationship` usage.

Most users should not need code changes. One edge case remains: nullable (`Optional[...]`) fields may accept omitted values as `None` where strict “required but nullable” validation was previously relied on.

Added a short smoke-test of `gqlalchemy` in macOS and restored tests for Windows in CI.

### Pull request type

Please delete options that are not relevant.

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring with functional or API changes
- [ ] Refactoring without functional or API changes
- [x] Build or packaging related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

### Related issues

Delete section if this PR doesn't resolve any issues. 

Closes (link to issue)

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

######################################

### Reviewer checklist (the reviewer checks this part)
- [ ] Core feature implementation
- [ ] Tests
- [ ] Code documentation
- [ ] Documentation on [gqlalchemy/docs](https://github.com/memgraph/gqlalchemy/tree/main/docs)

######################################

### CODE REVIEW

File: tests/docs/test_ogm.py [hunk 1]
```diff
@@ -13,10 +13,19 @@
...
+import pytest
 
 from gqlalchemy import Memgraph, Node, Relationship, Field, match
 from gqlalchemy.query_builders.memgraph_query_builder import Operator
 
+if sys.version_info >= (3, 14):
+    pytest.skip(
+        "OGM docs tests rely on Pydantic v1 behavior not compatible with Python 3.14+.",
+        allow_module_level=True,
+    )
```
[the remark this answers is not in the document]
Comment (mattkjames7) [reply tests/docs/test_ogm.py]: Actually, I should re-enable this, given that I have updated the code to use pydantic v2.

File: pyproject.toml [hunk 2]
```diff
@@ -19,26 +19,31 @@ dependencies = [
     "adlfs>=2023.9.0,<2025.0.0",
     "neo4j>=4.4.3,<6.0.0",
     "numpy>=1.26.4,<3.0",
-    "pandas<=2.2.3",
-    "tensorflow==2.16.2",
+    "pandas<=2.2.3; python_version >= '3.10' and python_version < '3.14'",
+    "pandas>=2.3,<4.0; python_version >= '3.14' and python_version < '4.0'",
+    "tensorflow==2.16.2; (python_version >= '3.10' and python_version < '3.12' and sys_platform != 'win32') or (python_version >= '3.12' and python_version < '3.13')",
 ]
```
[the remark this answers is not in the document]
Comment (mattkjames7) [reply pyproject.toml]: This is intended - Windows has binaries for tensorflow on Python 3.12.

File: pyproject.toml [hunk 3]
```diff
@@ -19,26 +19,31 @@ dependencies = [
...
+    "pyarrow>=22,<24; python_version >= '3.13' and python_version < '4.0'",
+]
 tfgnn = [
-    "tensorflow-gnn>=1.0.0,<2.0.0",
-    "tensorflow==2.16.2",
-    "tf-keras>=2.16,<3.0",
+    "tensorflow-gnn>=1.0.0,<2.0.0; (python_version >= '3.10' and python_version < '3.12' and sys_platform != 'win32') or (python_version >= '3.12' and python_version < '3.13')",
+    "tensorflow==2.16.2; (python_version >= '3.10' and python_version < '3.12' and sys_platform != 'win32') or (python_version >= '3.12' and python_version < '3.13')",
+    "tf-keras>=2.16,<3.0; (python_version >= '3.10' and python_version < '3.12' and sys_platform != 'win32') or (python_version >= '3.12' and python_version < '3.13')",
 ]
```
[the remark this answers is not in the document]
Comment (mattkjames7) [reply pyproject.toml:37]: same issue as above

File: .github/workflows/build-and-test.yml [hunk 4]
```diff
@@ -12,40 +12,81 @@ on:
...
+        uses: actions/setup-python@a309ff8b426b58ec0e2a45f0f869d46889d02405 # v6.2.0
         with:
           python-version: ${{ matrix.python-version }}
 
+      - name: Set up pip and install packages
+        run: |
+          # these tools installed globally for Memgraph
+          python3 -m pip install --upgrade pip setuptools wheel
+          sudo -H python3 -m pip install networkx numpy scipy
+
```
Comment (antejavor) [inline .github/workflows/build-and-test.yml]: Should this be now a `uv`? Since we are setting up a `uv` but noting major.

File: .github/workflows/build-and-test.yml [hunk 5]
```diff
@@ -61,64 +102,132 @@ jobs:
...
+            echo "Skipping vm.max_map_count update: /proc/sys/vm/max_map_count not available in this WSL environment."
+          fi
+
       - name: Download, install and run Memgraph under WSL
         shell: wsl-bash {0} # root shell
         run: |
           sudo apt-get update
           sudo apt-get -y install python3 python3-pip ipython3
-          pip3 install networkx numpy scipy
+          pip3 install networkx numpy scipy --break-system-packages
```
Comment (antejavor) [inline .github/workflows/build-and-test.yml:189]: Making Win work is already a bonus, also possibly a `uv`, but feel free to ignore for now.

File: .github/workflows/build-and-test.yml [hunk 6]
```diff
@@ -134,33 +243,73 @@ jobs:
...
-        uses: actions/upload-artifact@v4
+        uses: actions/upload-artifact@bbbca2ddaa5d8feaa63e36b76fdaad77386f024f # v7.0.0
         if: always()
         with:
-          name: neo4j-log
+          name: neo4j-logs-${{ env.OS }}-py${{ matrix.python-version }}
           path: C:\neo4j
           overwrite: true
+
+  build_and_test_macos:
```
Comment (antejavor) [inline .github/workflows/build-and-test.yml:288]: It is art to watch this setup compared to windows :D ahh...

**Review by antejavor — APPROVED** [review]
Massive improvements all around the places (python support, win, macos build) , with a minor downside in compatibility. 

I guess at some point we will ditch most of the dependencies out and move to Memgraph QM ->GNN/TGNN, being an example, since we now have this ->https://github.com/memgraph/memgraph/blob/master/mage/python/gnn.py 

That means dependencies can live inside the mage container, keeping clients cleaner and leaner.  🦾

File: .github/workflows/build-and-test.yml (hunk 4 above)
Comment (mattkjames7) [reply .github/workflows/build-and-test.yml]: This needs to exist on the system, I'll check if `uv` works (I wasn't sure whether `uv` could install globally, but I Googled and it looks like it should).

File: .github/workflows/build-and-test.yml (hunk 6 above)
Comment (mattkjames7) [reply .github/workflows/build-and-test.yml:288]: :joy: Windows is nightmare-level for CI

File: .github/workflows/build-and-test.yml (hunk 5 above)
Comment (mattkjames7) [reply .github/workflows/build-and-test.yml:189]: I'll leave this as a pip install for now (I tried uv, but it looks like the uv action doesn't install for WSL).

**Review by antejavor — APPROVED** [review]

**Comment by Jdyn** [comment]
Hey, is this good to go for a new release?
